### PR TITLE
Using pinched grid for numerical aquifer

### PR DESCRIPTION
**Pull request by GitPaean**
Downstream PRs are https://github.com/OPM/opm-grid/pull/517 and https://github.com/OPM/opm-simulators/pull/3134 and OPM/opm-upscaling#331

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

File: src/opm/parser/eclipse/EclipseState/Aquifer/Aquancon.cpp [hunk 1]
```diff
@@ -108,6 +108,7 @@ namespace Opm {
 
     Aquancon::Aquancon(const EclipseGrid& grid, const Deck& deck)
     {
+        const std::vector<int>& actnum = grid.getACTNUM();
```
Comment (GitPaean) [inline src/opm/parser/eclipse/EclipseState/Aquifer/Aquancon.cpp:111]: From here, the using of pinched grid for connection generation, will not affect the analytical aquifers. This PR is for numerical aquifers, the analytical aquifers will be handled in a separate PRs for the ease of testing and analysis.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
It looks triggering jenkins test from upstream module will not guarantee the latest code will be used downstream.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by joakim-hove** [comment]
>It looks triggering jenkins test from upstream module will not guarantee the latest code will be used downstream.

Yes - that is frustrating. It sometimes helps to close and reopen downstream

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
This PR is basically trying to use processed grid (opm-grid) to generate the connections for numerical aquifers. The main difficulties are in the opm-grid module. Numerical aquifers appear to be very invading to the grid processing code, but unless we want to do the grid processing (determination of the active cells) in the parser (at the early stage of the parsing), it will be invading.  The work in opm-common is mostly about providing interfaces.  I just marked this PR to be ready for review. I will also do so for the downstream PRs.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
> Yes - that is frustrating. It sometimes helps to close and reopen downstream

It looks like this did help for this case.  Previously, it got stuck no matter what I do.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 please

**Comment by GitPaean** [comment]
Interface in opm-upscaling needs to be updated. Will do.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
It is very weird the following tests failed, while other numerical aquifer cases (including the field case) are not affected at all. Will investigate. 

```
mpi.compareECLFiles_flow+3D_2AQU_NUM
mpi.compareParallelSim_flow+3D_2AQU_NUM
```

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
> It is very weird the following tests failed, while other numerical aquifer cases (including the field case) are not affected at all. Will investigate.

```
mpi.compareECLFiles_flow+3D_2AQU_NUM
mpi.compareParallelSim_flow+3D_2AQU_NUM
```

It is really tricky and as long as we have connection with K of 9 or 10 (K dimension is 10), this PR causes difference in the results.

**Comment by GitPaean** [comment]
> It is really tricky and as long as we have connection with K of 9 or 10 (K dimension is 10), this PR causes difference in the results.

At the end, it is the direction of NNC playing some role here. New commits make sure the NNC is from `low` to `high` for NNCs generated by numerical aquifers.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
Now only jenkins test failures are in opm-upscaling. Will try to fix them now.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
Finally the jenkins is green. I will request reviews now.

**Comment by totto82** [comment]
@GitPaean Do we want this in the release branch? If so please tag the relevant branches.

**Comment by GitPaean** [comment]
> @GitPaean Do we want this in the release branch? If so please tag the relevant branches.

Sure, why not. If it happens to take too long, we can always remove it later.

File: opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/SingleNumericalAquifer.hpp [hunk 2]
```diff
@@ -57,7 +57,9 @@ namespace Opm {
 
         std::unordered_map<size_t, AquiferCellProps> aquiferCellProps() const;
 
-        std::vector<NNCdata> aquiferNNCs(const EclipseGrid& grid, const FieldPropsManager& fp) const;
+        std::vector<NNCdata> aquiferCellNNCs() const;
+        std::vector<NNCdata> aquiferConnectionNNCs(const EclipseGrid &grid, const FieldPropsManager &fp) const;
```
Comment (GitPaean) [inline opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/SingleNumericalAquifer.hpp:61]: FYI. For NNCs from numerical aquifers, it can be between the aquifer cells, and also between aquifer cell and aquifer connections. For the one between aquifer cells, we can generate on-site, for the one between aquifer cell and aquifer connections, we do it during the grid processing in opm-grid. That is why we split the `aquifernNNCs` to be two separate functions here.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
@joakim-hove, will you be able to review this PR soon? Thanks. If possible, we would like to get this work in so the release branch can be generated.

**Comment by joakim-hove** [comment]
> will you be able to review this PR soon? 

yes - to be honest I thought I had done it; but I see now that the comments were only in my head. I'll call you

**Comment by joakim-hove** [comment]
This is complicated and I must admit I am struggling to get a proper understanding/overview of the changes. I don't think we should rush this to get it in for the release, or hold the release for it.

**Comment by totto82** [comment]
Are these PRs crucial to have a working numerical aquifer code?

**Comment by totto82** [comment]
I will continue the release process without it.

**Comment by GitPaean** [comment]
> I will continue the release process without it.

I will remove the release tag from it.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by GitPaean** [comment]
jenkins is green now. Please continue the review process. Thanks.

File: opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquifers.hpp [hunk 3]
```diff
@@ -46,10 +46,14 @@ namespace Opm {
 
         std::unordered_map<size_t, double> aquiferCellVolumes() const;
 
-        std::vector<NNCdata> aquiferNNCs(const EclipseGrid& grid, const FieldPropsManager& fp) const;
+        std::vector<NNCdata> aquiferCellNNCs() const;
+        std::vector<NNCdata> aquiferConnectionNNCs(const EclipseGrid& grid, const FieldPropsManager& fp) const;
 
         std::unordered_map<size_t, AquiferCellProps> aquiferCellProps() const;
 
+        void preGeneratingConnections(const Deck& deck, const EclipseGrid& grid);
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquifers.hpp:54]: These functions should not have the `xxxing` form in the name - rather:
```
void preGenerateConnections(); // or initConnections()?
void postProcessConnections();
```
since this is so complex it is worth it to fiddle with the names - sorry.

**Review by joakim-hove — APPROVED** [review]
With a very minor function name change I am ready to merge this. What an effort!

**Comment by GitPaean** [comment]
jenkins build this opm-grid=517 opm-simulators=3134 opm-upscaling=331 please

**Comment by joakim-hove** [comment]
OK - this has been a consistent effort! I'll merge the family now.

**Comment by GitPaean** [comment]
Thanks for reviewing and merging it. There might be more fixing and improvements coming, but at least the relatively complete implementation is in place now.

**Comment by bska** [comment]
I think series of PR may have [broken the sequential build](https://ci.opm-project.org/job/opm-simulators/781/consoleFull).  Please investigate and fix.

**Comment by GitPaean** [comment]
> I think series of PRs may have broken the sequential build. Please investigate and fix.

Sure. I always have MPI on so it is very possible.
